src/main/java/org/assertj/core/api/AbstractBufferAssert.java

```diff
@@ -0,0 +1,189 @@
+import static org.assertj.core.error.buffer.ShouldHaveLength.shouldHaveLength;
+import static org.assertj.core.error.buffer.ShouldHaveRemainingLength.shouldHaveRemainingLength;
+
+import java.nio.Buffer;
+
+/**
+ * Base class for all implementations of assertions for {@link Buffer}s.
+ *
+ * @param <SELF> the "self" type of this assertion class. Please read &quot;<a href="http://bit.ly/1IZIRcY"
+ *          target="_blank">Emulating 'self types' using Java Generics to simplify fluent API implementation</a>&quot;
+ *          for more details.
+ * @param <ACTUAL> the type of the "actual" buffer.
+ *
+ * @author Jean de Leeuw
+ */
+public abstract class AbstractBufferAssert<SELF extends AbstractBufferAssert<SELF, ACTUAL>, ACTUAL extends Buffer> extends AbstractAssert<SELF, ACTUAL> {
+
+  public AbstractBufferAssert(ACTUAL actual, Class<?> selfType) {
+    super(actual, selfType);
+  }
+
+  /**
+   * Verifies that the actual {@code Buffer} has been flipped.
+   *
+   * <b>Important:</b>
+   * Due to the limitations of the Buffer API there is, short of looping over the buffer and checking the content one
+   * by one, no way to check if the Buffer is actually flipped. It is impossible to make a distinction between
+   * a buffer that is filled to its capacity and a buffer that simply has not been flipped (or manual repositioning
+   * of the buffer). Due to these limitations, this method only checks if the position of the buffer equals 0.
+   * <p>
+   * Example:
+   *
+   * <pre><code class='java'>
+   * Buffer buffer = ByteBuffer.allocate(10);
+   *
+   * // this assertion succeeds ...
+   * buffer.flip();
+   * assertThat(buffer).isFlipped();
+   *
+   * // ... but this one fails as "allocate" does not flip the buffer.
+   * assertThat(buffer).isFlipped();
+   * </code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual {@code Buffer} is {@code null}.
+   * @throws AssertionError if the actual {@code Buffer} has not been flipped.
+   */
+  public SELF isFlipped() {
+    isNotNull();
+    if (actual.position() != 0) throwAssertionError(shouldBeFlipped(actual));
+    return myself;
+  }
+
+  /**
+   * Verifies that the actual {@code Buffer} is empty.
+   *
+   * Example:
+   *
+   * <pre><code class='java'>
+   * // this assertion succeeds ...
+   * Buffer buffer = ByteBuffer.wrap("".getBytes());
+   * assertThat(buffer).isEmpty();
+   *
+   * // ... but this one fails as "buffer" is not empty.
+   * Buffer buffer = ByteBuffer.wrap("test".getBytes());
+   * assertThat(buffer).isEmpty();
+   * </code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual {@code Buffer} is {@code null}.
+   * @throws AssertionError if the actual {@code Buffer} has not been flipped.
+   * @throws AssertionError if the actual {@code Buffer} is not empty.
+   */
+  public SELF isEmpty() {
+    isNotNull();
+    isFlipped();
```

Comment:
Just curious why is it needed ? 
it it is it should made it clear the javadoc and have a failing example with an unflipped buffer.

Reply:
ok, after some googling I guess it is required to read data.

Reply:
Well the problem is that it is not exactly required (as in you could read data from the buffer without flipping first), but it is more of a convention to properly distinguish between writing to the buffer and reading from the buffer.

I decided to "enforce" this convention by doing the isFlipped check on all operations (which does a very basic check to determine if the buffer is flipped or not, but cannot actually guarantee it as explained in the "Buffer Flips" section of this PR), because I assume that if you are going to perform assertions on your buffer, that you have finished writing to the buffer.

I am not a heavy user of buffers however, so there might be a valid use case where you do not want to flip the buffer before performing assertions on it, proving my assumption wrong. In which case we should perhaps reconsider this isFlipped check.